sdk/core/azure-core/inc/azure/core/internal/json/json_optional.hpp

```diff
@@ -37,11 +37,24 @@ namespace Azure { namespace Core { namespace Json { namespace _internal {
     static inline void SetIfExists(
         Azure::Nullable<T>& destination,
         Azure::Core::Json::_internal::json const& jsonKey,
-        std::string const& key) noexcept
+        std::string const& key)
     {
       if (jsonKey.contains(key) && !jsonKey[key].is_null()) // In Json and not-Null
       {
-        destination = jsonKey[key].get<T>();
+        try
+        {
+          destination = jsonKey[key].get<T>();
+        }
+        catch (std::exception const& e)
+        {
+          // Catch the Azure::Core::Json::_internal::detail::type_error thrown by the 3rd party
+          // library which we don't want to expose.
+          throw std::runtime_error(
+              std::string(
+                  "Could not find required field '" + key + "' of type '" + typeid(T).name()
```

Comment:
As far as I know, the value of `typeid(T)` is compiler specific and the value of it varies between compilers. However I'm also wondering if it works if RTTI is disabled.

Reply:
Although this may get resolved at compile-time, we can't expect this to work when RTTI is disabled. We need an
```
#include <azure/core/rtti.hpp>
...
"Could not find required field '" + key + "'"
#ifndef(AZ_CORE_RTTI)
" of type '" + typeid(T).name() + "'"
#endif
"."
```

Reply:
I wonder what happens when RTTI is disabled. Does it return an empty string?
Do we have any CI legs with that option? I see these in `platform-matrix-live`.
https://github.com/Azure/azure-sdk-for-cpp/blob/7da5a4568c0c809bd05489a786c3b77f146d601e/eng/pipelines/templates/stages/platform-matrix-live.json#L61-L63

We don't have any `-DBUILD_RTTI=OFF` in `platform-matrix`.

Reply:
Or, you could write a template with specializations for all the types that we might be getting the value for - I bet the `bool`, `float`, `double`, `char`, `std::int8_t`, `std::uint8_t`, `std::int16_t`, `std::uint16_t`, `std::int32_t`, `std::uint32_t`, `std::int64_t`, `std::uint64_t`, and `std::string` will cover everything that is currently used and may potentially be used.
```cpp
template<typename>
constexpr const char* GetTypeName();
// never define this so that if someone uses it with an unknown type,
// it does not compile. Only define the template specializations:

template<>
constexpr const char* GetTypeName<bool>() { return "bool"; }

template<>
constexpr const char* GetTypeName<std::string>() { return "std::string"; }

// Usage:: "can't get property for type '" + GetTypeName<T>() + "'."
```

Reply:
Yep, looks like non-MSVC compilers return mangled names of types, like `NSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEE` or `b` for bool:
https://dev.azure.com/azure-sdk/29ec6040-b234-4e31-b139-33dc4287b756/_apis/build/builds/2906396/logs/744